agora/Agora/Effect/TreasuryWithdrawal.hs

```diff
@@ -0,0 +1,153 @@
+    receivers <- plet $ pfromData $ pfield @"receivers" # datum'
+    txInfo <- pletFields @'["outputs", "inputs"] txInfo'
+    PJust txOut <- pmatch $ findTxOutByTxOutRef # txOutRef' # pfromData txInfo'
+    effInput <- pletFields @'["address", "value"] $ txOut
+    let outputValues =
+          pmap
+            # plam
+              ( \(pfromData -> out') -> P.do
+                  out <- pletFields @'["address", "value"] $ out'
+                  cred <- pletFields @'["credential"] $ pfromData out.address
+                  pdata $ ptuple # cred.credential # out.value
+              )
+            # txInfo.outputs
+        inputValues =
+          pmap
+            # plam
+              ( \((pfield @"resolved" #) . pfromData -> txOut') -> P.do
+                  txOut <- pletFields @'["address", "value"] $ txOut'
+                  pdata $ ptuple # txOut.address # txOut.value
+              )
+            # txInfo.inputs
+        treasuryInputValues =
+          pfilter
+            # plam (\((pfield @"_0" #) . pfromData -> addr) -> pnot #$ addr #== effInput.address)
+            # inputValues
+        treasuryCredentials =
+          pmap
+            # plam ((pfield @"credential" #) . pfromData . (pfield @"_0" #) . pfromData)
+            # treasuryInputValues
+        treasuryOutputValues =
+          pfilter
+            # plam
+              ( \((pfield @"_0" #) . pfromData -> addr) -> P.do
+                  pelem # addr # treasuryCredentials
+              )
+            # outputValues
+        treasuryInputValuesSum =
+          pfoldr
+            # plam (\((pfield @"_1" #) . pfromData -> x) y -> paddValue # pfromData x # y)
+            # pconstant (mempty :: Value)
+            # treasuryInputValues
+        treasuryOutputValuesSum =
+          pfoldr
+            # plam (\((pfield @"_1" #) . pfromData -> x) y -> paddValue # pfromData x # y)
+            # pconstant (mempty :: Value)
+            # treasuryOutputValues
+        receiverValuesSum =
+          pfoldr
+            # plam (\((pfield @"_1" #) . pfromData -> x) y -> paddValue # pfromData x # y)
+            # pconstant (mempty :: Value)
+            # receivers
+        outputContentMatchesRecivers =
+          pall # plam (\out -> pelem # out # outputValues)
+            #$ receivers
+        excessShouldBePaidToInputs =
+          pdata (paddValue # receiverValuesSum # treasuryOutputValuesSum) #== pdata treasuryInputValuesSum
+        shouldNotPayToEffect =
+          pnot #$ pany
+            # plam
+              ( \x ->
+                  effInput.address #== pfield @"address" # pfromData x
+              )
+            # pfromData txInfo.outputs
+
+    passert "Transaction output does not match receivers" outputContentMatchesRecivers
+    passert "Transaction should not pay to effects" shouldNotPayToEffect
+    passert "Remainders should be returned to the treasury" excessShouldBePaidToInputs
```

Comment:
Alright, so basically these three checks are good. But this currently doesn't prevent misuse of the GAT by the one who signs this transaction. An example attack vector is:
- Correctly perform the actions as expected (i.e. payout from the treasury correctly)
- _Also_ mutate a protocol parameter, or mutate the governor datum to nuke the governor and prevent all proposals in the future. 

So, there's a number of ways it could be misused in the same transaction. You need to somehow also check that the only involved scripts are the ones you expected. This could be as simple as checking that all inputs & outputs are only the "known" credentials. e.g. "the set of all input and output credentials is the treasury, the receivers, the signer's pk". This will already prevent most issues.

Reply:
So do I need to add field like "treasuries" for the datum? So that it can explicitly state what treasury to use?

Reply:
Well, yeah you probably do now that I think of it :thinking:. You do need to specify the address of where it can spend from, that's right.